### PR TITLE
Add Basic navigator.mediaDevices test.

### PR DESCRIPTION
**Pull request by agouaillard**
=> DOM

### CODE REVIEW

File: mediacapture-streams/enumerating-local-media-devices/mediadevices/api-present.html [hunk 1]
```diff
@@ -0,0 +1,24 @@
+<!doctype html>
+<html>
+<head>
+<title>getUserMedia: test that getUserMedia is present (with or without vendor prefix)</title>
```
Comment (alvestrand) [inline mediacapture-streams/enumerating-local-media-devices/mediadevices/api-present.html]: Ths test checks that enumerateDevices is present, while the title, file name and so on claim to check for getUserMedia. Can you make it consistent?

File: mediacapture-streams/obtaining-local-multimedia-content/mediadevices/api-present.html [hunk 2]
```diff
@@ -0,0 +1,24 @@
+<!doctype html>
+<html>
+<head>
+<title>getUserMedia: test that getUserMedia is present (with or without vendor prefix)</title>
+<link rel="author" title="Dr Alex Gouaillard" href="mailto:agouaillard@gmail.com"/>
+<link rel="help" href="https://www.w3.org/TR/mediacapture-streams/#mediadevices-interface-extensions">
+<meta name='assert' content='Check that the mediaDevices.getUserMedia() method is present.'/>
+<meta name='flags' content='dom'/>
```
Comment (alvestrand) [inline mediacapture-streams/obtaining-local-multimedia-content/mediadevices/api-present.html]: What does this meta do?

File: mediacapture-streams/obtaining-local-multimedia-content/mediadevices/api-present.html [hunk 3]
```diff
@@ -0,0 +1,24 @@
+<!doctype html>
+<html>
```
Comment (alvestrand) [inline mediacapture-streams/obtaining-local-multimedia-content/mediadevices/api-present.html]: Why does this test exist twice in two different subdirs?

File: mediacapture-streams/obtaining-local-multimedia-content/navigatorusermedia/deny.html [hunk 4]
```diff
@@ -26,8 +26,22 @@ <h1 class="instructions">Description</h1>
...
   }),
   t.step_func(function (error) {
-    assert_equals(error.name, "SecurityError", "SecurityError returned");
-    assert_equals(error.constraintName, null, "constraintName attribute not set for permission denied");
+    // NOTE ALEX: last call specs and editor specs differ here:
+    // editor draft, 2016: https://w3c.github.io/mediacapture-main/getusermedia.html
+    // assert_equals(error.name, "securityError", "securityError returned");
+    // last call working draft, 2016: https://www.w3.org/TR/mediacapture-streams/
+    // assert_equals(error.name, "PermissionDeniedError", "PermissionDeniedError returned");
+    if(error.name != "SecurityError" &&          // FF
```
Comment (alvestrand) [inline mediacapture-streams/obtaining-local-multimedia-content/navigatorusermedia/deny.html]: Last Call spec is old. Follow the editors' draft.
Note that error codes is a part of the spec that has been in flux - but for the getusermedia spec, we hope it's stable now.

**Comment by alvestrand** [comment]
Seems to me you're doing an extremely deep hierarchy: mediacapture-streams/obtaining-local-multimedia-content/{navigatorusermedia,mediadevices}/<testname>.

Any reason for this deepness? If so, can you explain proper placement in a README.md file in the mediacapture-streams directory?
My immediate thought is that the "navigatorusermedia" level should go away.

File: mediacapture-streams/OWNERS [hunk 5]
```diff
@@ -1,3 +1,4 @@
 @alvestrand
 @dontcallmedom
 @foolip
+@agouaillard
```
Comment (foolip) [inline mediacapture-streams/OWNERS]: Can you keep the file sorted?

File: mediacapture-streams/obtaining-local-multimedia-content/navigatorusermedia/deny.html [hunk 6]
```diff
@@ -26,8 +26,22 @@ <h1 class="instructions">Description</h1>
...
+       error.name != "PermissionDeniedError") {  // all others
+      assert_unreached("Error type is neither securityError or PermissionDeniedError.");
+      t.done();
+    };
+    // NOTE ALEX: this is also nullable, but practically, all the browsers but FF set it to an empty string.
```
Comment (foolip) [inline mediacapture-streams/obtaining-local-multimedia-content/navigatorusermedia/deny.html]: File a spec bug or assert the behavior that the spec requires.

File: mediacapture-streams/obtaining-local-multimedia-content/navigatorusermedia/deny.html (hunk 4 above)
Comment (foolip) [inline mediacapture-streams/obtaining-local-multimedia-content/navigatorusermedia/deny.html]: Also update any links to https://w3c.github.io/* where that is the most up to date version.

File: mediacapture-streams/enumerating-local-media-devices/mediadevices/api-present.html [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+<link rel="author" title="Dr Alex Gouaillard" href="mailto:agouaillard@gmail.com"/>
+<link rel="help" href="https://www.w3.org/TR/mediacapture-streams/#mediadevices">
```
Comment (foolip) [inline mediacapture-streams/enumerating-local-media-devices/mediadevices/api-present.html]: Link to http://w3c.github.io/mediacapture-main/#enumerating-devices

File: mediacapture-streams/enumerating-local-media-devices/mediadevices/api-present.html [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+<h1 class="instructions">Description</h1>
+<p class="instructions">This test checks for the presence of the
+<code>navigator.mediaDevices.enumerateDevices()</code> method.</p>
+<div id='log'></div>
+<script src=/resources/testharness.js></script>
+<script src=/resources/testharnessreport.js></script>
+<script>
+test(function () {
+  assert_true(undefined !== navigator.mediaDevices.enumerateDevices(), "navigator.mediaDevices.enumerateDevices() exists");
+  }, "mediaDevices.enumerateDevices() is present on navigator");
```
Comment (foolip) [inline mediacapture-streams/enumerating-local-media-devices/mediadevices/api-present.html]: This is strangely described. It's `navigator.mediaDevices.enumerateDevices` that can be present, but what you're actually testing is that when invoked, it returns something other than undefined. Can you assert that it returns what the spec requires instead, i.e. a promise? AFAICT, the spec never rejects the promise, so you could even verify that it resolves to a `sequence<MediaDeviceInfo>`, even if empty.

File: mediacapture-streams/enumerating-local-media-devices/mediadevices/api-present.html [hunk 9]
```diff
@@ -0,0 +1,24 @@
+<!doctype html>
+<html>
+<head>
+<title>getUserMedia: test that getUserMedia is present (with or without vendor prefix)</title>
+<link rel="author" title="Dr Alex Gouaillard" href="mailto:agouaillard@gmail.com"/>
+<link rel="help" href="https://www.w3.org/TR/mediacapture-streams/#mediadevices">
+<meta name='assert' content='Check that the mediaDevices.enumerateDevices() method is present.'/>
+<meta name='flags' content='dom'/>
+<link rel='stylesheet' href='/resources/testharness.css' media='all'/>
```
Comment (foolip) [inline mediacapture-streams/enumerating-local-media-devices/mediadevices/api-present.html]: Won't this be included automatically? In general, this test is far more verbose than it needs to be, a lot of this is added automatically. In its current form, it could be reduced to
```HTML
<!doctype html>
<title>mediaDevices.enumerateDevices() returns a promise</title>
<script src=/resources/testharness.js></script>
<script src=/resources/testharnessreport.js></script>
<script>
assert_true(navigator.mediaDevices.enumerateDevices() instanceof Promise);
</script>
```

I'll defer to @hta, but at least I much prefer minimizing tests as much as possible so that the interesting bits stand out.

File: mediacapture-streams/enumerating-local-media-devices/mediadevices/api-present.html (hunk 1 above)
Comment (agouaillard) [inline mediacapture-streams/enumerating-local-media-devices/mediadevices/api-present.html]: will do.

File: mediacapture-streams/obtaining-local-multimedia-content/mediadevices/api-present.html (hunk 2 above)
Comment (agouaillard) [inline mediacapture-streams/obtaining-local-multimedia-content/mediadevices/api-present.html]: which one?

File: mediacapture-streams/obtaining-local-multimedia-content/mediadevices/api-present.html (hunk 3 above)
Comment (agouaillard) [inline mediacapture-streams/obtaining-local-multimedia-content/mediadevices/api-present.html]: After discussion with dom, we decided to follow the spec chapter layout. I did that exactly.

File: mediacapture-streams/obtaining-local-multimedia-content/navigatorusermedia/deny.html (hunk 4 above)
Comment (agouaillard) [inline mediacapture-streams/obtaining-local-multimedia-content/navigatorusermedia/deny.html]: yes, will do.

File: mediacapture-streams/OWNERS (hunk 5 above)
Comment (agouaillard) [inline mediacapture-streams/OWNERS]: @alvestrand : we decided with dom to follow the chapters in the spec for easier spec <-> tests scrum'ing.

@foolip will do.

File: mediacapture-streams/obtaining-local-multimedia-content/navigatorusermedia/deny.html (hunk 6 above)
Comment (agouaillard) [inline mediacapture-streams/obtaining-local-multimedia-content/navigatorusermedia/deny.html]: that has been resolved of line. My mistake the original test was actually correct. I m correcting that.

File: mediacapture-streams/enumerating-local-media-devices/mediadevices/api-present.html (hunk 9 above)
Comment (agouaillard) [inline mediacapture-streams/enumerating-local-media-devices/mediadevices/api-present.html]: I just adapted the existing test.
happy either way.
can someone rule on this?

File: mediacapture-streams/obtaining-local-multimedia-content/navigatorusermedia/deny.html [hunk 10]
```diff
@@ -26,8 +26,22 @@ <h1 class="instructions">Description</h1>
...
   t.step_func(function (error) {
-    assert_equals(error.name, "SecurityError", "SecurityError returned");
-    assert_equals(error.constraintName, null, "constraintName attribute not set for permission denied");
+    // NOTE ALEX: last call specs and editor specs differ here:
+    // editor draft, 2016: https://w3c.github.io/mediacapture-main/getusermedia.html
+    // assert_equals(error.name, "securityError", "securityError returned");
+    // last call working draft, 2016: https://www.w3.org/TR/mediacapture-streams/
+    // assert_equals(error.name, "PermissionDeniedError", "PermissionDeniedError returned");
+    if(error.name != "SecurityError" &&          // FF
+       error.name != "PermissionDeniedError") {  // all others
```
Comment (agouaillard) [inline mediacapture-streams/obtaining-local-multimedia-content/navigatorusermedia/deny.html]: you re right, it was correct, i got confused between versions of the specs. correcting it right now.

File: mediacapture-streams/obtaining-local-multimedia-content/mediadevices/api-present.html (hunk 2 above)
Comment (alvestrand) [inline mediacapture-streams/obtaining-local-multimedia-content/mediadevices/api-present.html]: All of them, really, but the <meta name='flags' content='dom*/> in particular. Minimization is good.

File: mediacapture-streams/obtaining-local-multimedia-content/mediadevices/api-present.html (hunk 3 above)
Comment (alvestrand) [inline mediacapture-streams/obtaining-local-multimedia-content/mediadevices/api-present.html]: @dontcallmedom can you defend this practice, please? I regard our chapter structure as reasonably irrelevant to good test grouping.

File: mediacapture-streams/obtaining-local-multimedia-content/mediadevices/api-present.html (hunk 2 above)
Comment (agouaillard) [inline mediacapture-streams/obtaining-local-multimedia-content/mediadevices/api-present.html]: ok, i m removing them

File: mediacapture-streams/obtaining-local-multimedia-content/mediadevices/api-present.html (hunk 3 above)
Comment (agouaillard) [inline mediacapture-streams/obtaining-local-multimedia-content/mediadevices/api-present.html]: i already simplified it following your comment. cheers.

File: mediacapture-streams/enumerating-local-media-devices/api-present.html [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+<h1 class="instructions">Description</h1>
+<p class="instructions">This test checks for the presence of the
+<code>navigator.mediaDevices.enumerateDevices()</code> method.</p>
+<div id='log'></div>
+<script src=/resources/testharness.js></script>
+<script src=/resources/testharnessreport.js></script>
+<script>
+"use strict";
+//NOTE ALEX: for completion, a test for ondevicechange event is missing.
+test(function () {
```
Comment (alvestrand) [inline mediacapture-streams/enumerating-local-media-devices/api-present.html]: No objection to having this kind of test, but isn't this (API presence) better done by adding the WebIDL-driven tests that we did earlier for peerconnection API?

(to be clear: I'm happy merging this version, but @agouaillard should look at the WebIDL-driven tools)

File: mediacapture-streams/obtaining-local-multimedia-content/mediadevices/api-present.html (hunk 2 above)
Comment (alvestrand) [inline mediacapture-streams/obtaining-local-multimedia-content/mediadevices/api-present.html]: I don't see it gone...

**Comment by alvestrand** [comment]
I think we should merge this and progress iteratively. @dontcallmedom can decide.

File: mediacapture-streams/enumerating-local-media-devices/api-present.html [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+  assert_true(undefined !== navigator.mediaDevices.enumerateDevices(), "navigator.mediaDevices.enumerateDevices() exists");
```
Comment (dontcallmedom) [inline mediacapture-streams/enumerating-local-media-devices/api-present.html]: I think @foolip mentioned it before, but it seems to me more logical to test 
```js
undefined !== navigator.mediaDevices.enumerateDevices
```
since we're checking for existence

File: mediacapture-streams/enumerating-local-media-devices/api-present.html [hunk 13]
```diff
@@ -0,0 +1,48 @@
+<!doctype html>
+<html>
+<head>
+<title>enumerateDevices: test that enumerateDevices is present</title>
+<link rel="author" title="Dr Alex Gouaillard" href="mailto:agouaillard@gmail.com"/>
+<link rel="help" href="https://w3c.github.io/mediacapture-main/#enumerating-devices">
+<meta name='assert' content='Check that the enumerateDevices() method is present.'/>
+<link rel='stylesheet' href='/resources/testharness.css' media='all'/>
```
Comment (dontcallmedom) [inline mediacapture-streams/enumerating-local-media-devices/api-present.html]: I understand that we don't need (and thus shouldn't) link to the stylesheet, since it's added automatically by testharness.js.

File: mediacapture-streams/obtaining-local-multimedia-content/navigatorusermedia/deny.html [hunk 14]
```diff
@@ -26,8 +26,13 @@ <h1 class="instructions">Description</h1>
     t.done();
   }),
   t.step_func(function (error) {
-    assert_equals(error.name, "SecurityError", "SecurityError returned");
-    assert_equals(error.constraintName, null, "constraintName attribute not set for permission denied");
+    assert_equals(error.name, "securityError", "securityError returned");
+    // NOTE ALEX: this is also nullable, but practically, all the browsers but FF set it to an empty string.
```
Comment (dontcallmedom) [inline mediacapture-streams/obtaining-local-multimedia-content/navigatorusermedia/deny.html]: we should test based on what the spec says, not based on what implementations do; if and when the spec gets proven to be wrong, we will update the test, but if we hide implementation bugs in our test cases, we won't find the discrepancies between spec and implementations.

File: mediacapture-streams/obtaining-local-multimedia-content/mediadevices/api-present.html (hunk 2 above)
Comment (agouaillard) [inline mediacapture-streams/obtaining-local-multimedia-content/mediadevices/api-present.html]: I confirm it's gone in https://github.com/w3c/web-platform-tests/pull/2657/commits/a1aa788cf167b726deb48e47cbd6fb71568e7647

File: mediacapture-streams/enumerating-local-media-devices/api-present.html (hunk 11 above)
Comment (agouaillard) [inline mediacapture-streams/enumerating-local-media-devices/api-present.html]: You mean like the one used for media track API in mediastream-api/mediastreamtrack-init ?

File: mediacapture-streams/enumerating-local-media-devices/api-present.html (hunk 12 above)
Comment (agouaillard) [inline mediacapture-streams/enumerating-local-media-devices/api-present.html]: You mean without the parenthesis? Please educate me on the difference in JS.

File: mediacapture-streams/enumerating-local-media-devices/api-present.html (hunk 13 above)
Comment (agouaillard) [inline mediacapture-streams/enumerating-local-media-devices/api-present.html]: shall i modify them all then?

grep -R link * | grep testharness
enumerating-local-media-devices/api-present.html:<link rel='stylesheet' href='/resources/testharness.css' media='all'/>
mediastreams-as-media-elements/video-assignment-manual.html:<link rel='stylesheet' href='/resources/testharness.css' media='all'/>
obtaining-local-multimedia-content/mediadevices/api-present.html:<link rel='stylesheet' href='/resources/testharness.css' media='all'/>
obtaining-local-multimedia-content/navigatorusermedia/api-present.html:<link rel='stylesheet' href='/resources/testharness.css' media='all'/>
obtaining-local-multimedia-content/navigatorusermedia/deny.html:<link rel='stylesheet' href='/resources/testharness.css' media='all'/>
obtaining-local-multimedia-content/navigatorusermedia/empty-option-param.html:<link rel='stylesheet' href='/resources/testharness.css' media='all'/>
obtaining-local-multimedia-content/navigatorusermedia/getusermedia-impossible-constraint.html:<link rel='stylesheet' href='/resources/testharness.css' media='all'/>
obtaining-local-multimedia-content/navigatorusermedia/getusermedia-optional-constraint.html:<link rel='stylesheet' href='/resources/testharness.css' media='all'/>
obtaining-local-multimedia-content/navigatorusermedia/getusermedia-trivial-constraint.html:<link rel='stylesheet' href='/resources/testharness.css' media='all'/>
obtaining-local-multimedia-content/navigatorusermedia/unknownkey-option-param.html:<link rel='stylesheet' href='/resources/testharness.css' media='all'/>
stream-api/introduction/disabled-audio-silence.html:<link rel='stylesheet' href='/resources/testharness.css' media='all'/>
stream-api/introduction/disabled-video-black.html:<link rel='stylesheet' href='/resources/testharness.css' media='all'/>
stream-api/mediastream/audio.html:<link rel='stylesheet' href='/resources/testharness.css' media='all'/>
stream-api/mediastream/mediastream-addtrack.html:<link rel='stylesheet' href='/resources/testharness.css' media='all'/>
stream-api/mediastream/mediastream-finished-add.html:<link rel='stylesheet' href='/resources/testharness.css' media='all'/>
stream-api/mediastream/mediastream-gettrackid.html:<link rel='stylesheet' href='/resources/testharness.css' media='all'/>
stream-api/mediastream/mediastream-id-manual.html:<link rel='stylesheet' href='/resources/testharness.css' media='all'/>
stream-api/mediastream/mediastream-idl.html:<link rel='stylesheet' href='/resources/testharness.css' media='all'/>
stream-api/mediastream/mediastream-removetrack.html:<link rel='stylesheet' href='/resources/testharness.css' media='all'/>
stream-api/mediastream/stream-ended.html:<link rel='stylesheet' href='/resources/testharness.css' media='all'/>
stream-api/mediastream/video.html:<link rel='stylesheet' href='/resources/testharness.css' media='all'/>
stream-api/mediastreamtrack/mediastreamtrack-end.html:<link rel='stylesheet' href='/resources/testharness.css' media='all'/>
stream-api/mediastreamtrack/mediastreamtrack-id.html:<link rel='stylesheet' href='/resources/testharness.css' media='all'/>
stream-api/mediastreamtrack/mediastreamtrack-init.html:<link rel='stylesheet' href='/resources/testharness.css' media='all'/>
stream-api/video-and-audio-tracks/audiostreamtrack.html:<link rel='stylesheet' href='/resources/testharness.css' media='all'/>
stream-api/video-and-audio-tracks/videostreamtrack.html:<link rel='stylesheet' href='/resources/testharness.css' media='all'/>

File: mediacapture-streams/obtaining-local-multimedia-content/navigatorusermedia/deny.html (hunk 14 above)
Comment (agouaillard) [inline mediacapture-streams/obtaining-local-multimedia-content/navigatorusermedia/deny.html]: Agreed, and the commit you commented reverted to the original code that does just that as far as the error type is concerned. Now, some object are nullable by spec, but the spec does not address the default value at initialization. In that case, my understanding is that it is left to the implementation to decide if an empty string or a null object should be set. Do I understand wrong (and there is a spec that says that e.g. nullable object should be initialized as null)? Should we make the spec explicit in that case?

File: mediacapture-streams/enumerating-local-media-devices/api-present.html (hunk 12 above)
Comment (agouaillard) [inline mediacapture-streams/enumerating-local-media-devices/api-present.html]: ok, test existence, and not execute it. got it. fixing it.

File: mediacapture-streams/obtaining-local-multimedia-content/navigatorusermedia/deny.html (hunk 14 above)
Comment (dontcallmedom) [inline mediacapture-streams/obtaining-local-multimedia-content/navigatorusermedia/deny.html]: The only error objects that would have a non-undefined "constraintName" would be from overconstraint event; for any other error, the correct test is to `assert_equals(error.constraintName, undefined)`

File: mediacapture-streams/enumerating-local-media-devices/api-present.html (hunk 13 above)
Comment (dontcallmedom) [inline mediacapture-streams/enumerating-local-media-devices/api-present.html]: That'd be great (but preferably as a separate "clean up" pull request).

Comment (agouaillard) [inline mediacapture-streams/enumerating-local-media-devices/api-present.html]: will do in a separate PR.

File: mediacapture-streams/enumerating-local-media-devices/api-present.html (hunk 12 above)
Comment (agouaillard) [inline mediacapture-streams/enumerating-local-media-devices/api-present.html]: done.
